### ISSUE TITLE
Dismiss keyboard when pushing screen not working

### ISSUE THREAD

**Issue by jarrx** [wix/react-native-navigation#2318]
I want to dismiss the keyboard when pushing a new screen to navigator, but nothing happens and when I go back the keyboard is still open.
The container of the screen where the keyboard is, is a scrollview with `keyboardShouldPersistTaps='always'`.
So what I'm trying to do is
`Keyboard.dismiss()`
`this.props.navigator.push({
    screen: 'colorPicker',
    title: 'Color Picker'
})`

---
### Environment
* React Native Navigation version: ^1.1.300
* React Native version: 0.49.3
* Platform(s) (iOS, Android, or both?): iOS
* Device info (Simulator/Device? OS version? Debug/Release?): Simulator/Device. iOS version 11.0.3

**Comment by kjbappsllc** [comment]
Maybe you have found the fix to this, however, for future people looking at the same issue as I have been, here is how I fixed it:

`let dismissKeyBoard = () => {
    return new Promise((resolve, reject) => {
      Keyboard.dismiss();
      setTimeout(() => {
        resolve();
      }, 5);
    });
  };`

**Comment by mattmcdonald-uk** [comment]
I've also run into this issue, although when returning to the screen I'm unable to dismiss the keyboard at all. Definitely a bug here.

I'm also using a promise based workaround, though rather than an arbitrary time which may be too high or too low, no way of knowing, I'm waiting for the `keyboardWillChangeFrame` event which appears to be the first to occur when the keyboard dismisses. It doesn't seem to fire on Android though so I'm also using the `keyboardDidHide` event.

The promise will never resolve if the keyboard isn't open, so you'd need some way of tracking that to only use the method if needed.

````es6
new Promise((resolve, reject) => {
  const frameListener = Keyboard.addListener('keyboardWillChangeFrame', () => {
    frameListener.remove()
    didListener.remove()
    resolve()
  })

  const didListener = Keyboard.addListener('keyboardDidHide', () => {
    frameListener.remove()
    didListener.remove()
    resolve()
  })

  Keyboard.dismiss()
})
````

**Comment by jsina** [comment]
I've actually couldn't dismiss the keyboard after go back from pushed screen.

**Comment by raphaelhovsepyan** [comment]
Maybe this workaround can be helpful. Instead of using timeouts you can use `InteractionManager`.
```javascript
Keybaord.dismiss();
InteractionManager.runAfterInteractions(() => {
  // screen push logic
})
```

**Comment by zabojad** [comment]
It's a pity we are still having this issue with RNN `2.4.0`... Can we reopen it and treat it like a bug?

**Comment by huecnt-1493** [comment]
The issue still occure on IOS. I trying dimiss keyboard before push new screen but not working.
RN: 0.59.3

**Comment by oferRounds** [comment]
+1

**Comment by JerakRus** [comment]
+1

**Comment by chris-topher19** [comment]
+1, does anybody found any workaround? @huecnt-1493 @oferRounds @JerakRus

**Comment by oferRounds** [comment]
hi @chris-topher19 
The only workaround I found for now is to push the new screen after the keyboard dismissed. That is:
```
Keyboard.dismiss()
InteractionManager.runAfterInteractions(() => {
    this.props.navigator.push(navProps)
})
```

**Comment by byonghun** [comment]
After upgrading to `react-native@0.59.10`, the `InteractionManager.runAfterInteractions(() => {})` workaround does not work anymore. Anyone else running into the same issue after upgrading?

**Comment by byonghun** [comment]
@chris-topher19
Did the `InteractionManager.runAfterInteractions` work for you?

I was able to work around this issue with a really hacky way after I tried numerous ways to resolve the keyboard issue.

It's not the best way but it's something that works for now and something I will continue to monitor.

After trying to debug this issue, I stumbled on a way for the Keyboard to dismiss on Android.

Basically, showing an overlay before a push component action dismisses the Keyboard.

`InteractionManager.runAfterInteractions(async() => {
  Navigation.showOverlay({...})
  Navigation.push({...})
})`

**Comment by ayush-shta** [comment]
This workaround worked for me. Simply put the navigation push code inside a set timeout after a keyboard dismiss function call.

Keyboard.dismiss(); 
setTimeout(() => {
// Navigation push code
}, 0);

**Comment by hossamnasser938** [comment]
@byonghun  experiencing the same issue here with react-native@0.59.8

**Comment by hossamnasser938** [comment]
For me `requestIdleCallback` fixed the issue.
```js
Keyboard.dismiss();
 requestIdleCallback( () => {
          // push your screen here
} );
```

**Comment by ademcan** [comment]
Thanks @hossamnasser938 it worked for me. However, I'd be curious to know why it works if you have some more details :)

**Comment by noway** [comment]
still happening, none of the above works

**Comment by guyca** [comment]
@noway Can you please push a project with a reproduction?

**Comment by noway** [comment]
Sorry, don't think I have bandwidth for that... But what I found out which I hope helps someone: don't use Keyboard.dismiss and instead use `textInputRef.focus();textInputRef.blur()`, yep call focus first and then blur to actually dismiss the keyboard.

**Comment by rob5408** [comment]
> Maybe this workaround can be helpful. Instead of using timeouts you can use `InteractionManager`.
> 
> ```js
> Keybaord.dismiss();
> InteractionManager.runAfterInteractions(() => {
>   // screen push logic
> })
> ```

This also helps if you want to focus a text field when a functional component is first being presented (ie useEffect with an empty dependency list). Which technically will work to focus the text field, but breaks subsequent calls to `Keyboard.dismiss()`. The solution is to queue up your `ref.current.focus()` call inside the runAfterInteractions callback
